### PR TITLE
Paper: Combining Physics-Based and Data-Driven Modeling for Pressure Prediction in Well Construction

### PR DESCRIPTION
**Pull request by iodriller**
This is a pull request for the, SciPy Submission 52, paper by Oney Erge

### CODE REVIEW

**Comment by ivan-marroquin** [comment]
Hi @deniederhut ,

I will be glad to help with the review process of the abstract by @cbcunc on his work: Combining Physics-Based and Data-Driven Modeling for Pressure Prediction in Well Construction.

Kind regards,

ivan

**Comment by ucalyptus** [comment]
Hi I am Sayantan Das, one of the reviewers.Would take a look on this paper.

**Comment by ivan-marroquin** [comment]
Hi @ucalyptus ,

Sure. I will be glad to review the paper. How can I download the paper?

Ivan

**Comment by deniederhut** [comment]
The compiled PDF is available on our preprint server at procbuild.scipy.org. [Here is the direct link](http://procbuild.scipy.org/download/iodriller-2020_oney). You may find it easier to leave your comments inline on the source file, which [you can see here](https://github.com/scipy-conference/scipy_proceedings/pull/547/files?file-filters%5B%5D=.png&file-filters%5B%5D=.rst#diff-24a1f26052b27cff94aeb043c4c0cd68).

**Comment by ivan-marroquin** [comment]
Hi @deniederhut ,

OK. I just downloaded the paper.

Ivan

**Comment by deniederhut** [comment]
Hey @iodriller ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

**Comment by yingweiy** [comment]
Hi there, my name is Yingwei Yu, and I am happy to serve as reviewer for this paper. 

My comments are as follows. The paper is well-written in general. The authors give a very comprehensive literature review in the beginning part of the paper. The core method in this paper is described in section 3.3 and figure 8. From the raw inputs, multiple models, i.e., physic-based, NN-based, etc. are employed to predict SPP values, and a rule-based decision making algorithm and HMM make judgement and blend the predictive models results into the "final" output. Regarding the "rules" used, I am wondering how general they are. Are these rules general enough to be applied from stand to stand, and well to well? Do we still need to calibrate/redefine them just like calibrate the parameters of a physic model?  For example, Figure 10 and Table 1 show plausible results, are the training/parameters tuning done from other wells as a general model or based on the nearby stands from the same well? I suggest the authors to further elaborate how the HMM and the rule-based algorithm are defined in a little more detail.

**Comment by ivan-marroquin** [comment]
Hi @deniederhut ,

Find below my comments to authors:

Dear Authors,

I found your work very interesting. Indeed, there is a growing interest in finding ways to combine physics methods with data driven ones for better predictions. The authors showed that the proposed approach does not depend on a single machine learning technique, but different techniques can be tested. I believe this is an important step because users have the possibility to try out different approach and then, they can make better decisions. 

I have the following suggestions to improve the flow of discussion, refer to attached zip file:
- The authors should avoid using "this/these" to begin a sentence
- The authors should review text for minor grammar corrections  
- The authors should describe a bit better the results shown in all figures
- The authors should refer to figure 8 early in section 3.3
- The authors should refer to table 1 in all results were accuracy is discussed in sections 3.2 and 3.3
- In page 3, towards the end of last paragraph in section 3.2. The authors should replace "3" by "Fig. 3"

Last point. I do not think that the authors can claim that one machine learning technique is better than another based on results shown on table 1. The reason is simple. All these techniques were tested using different portions of the same input data. What would happen if the authors use the same training/test/validation data sets on all tested machine learning techniques, and the improvement in RMSE is not that much when going from a simple neural network to a more complicated gaussian process with markov chain. Why the gaussian process can be considered to be the best technique? How the authors would advice users to differentiate a desirable result from a result due to overfitting?

Kind regards,

Ivan

[Reviewed_Combining_Physics_DataDriven_Well_Construction.zip](https://github.com/scipy-conference/scipy_proceedings/files/4808084/Reviewed_Combining_Physics_DataDriven_Well_Construction.zip)

**Comment by iodriller** [comment]
Hi @ivan-marroquin,

Thank you very much for reviewing the manuscript. The authors appreciate your contribution to the work.  Please find the response to the comments below:

•	As the reviewer recommended, the manuscript is revised and “this/these” at the beginning of the sentences are revised.
•	The text is reviewed from the beginning to correct any grammar errors.
•	There are minor changes made at the captions at the text to accommodate this comment.
•	Figure 8th is referred to earlier in the revised version in section 3.3.
•	As the reviewer recommended, table 1 is referred to earlier as the accuracy of the models was mentioned.
•	As the reviewer pointed out, "3" is replaced by "Fig. 3".

Additionally:
•	The reviewer’s in-text corrections are implemented at the revised manuscript.
•	About reviewer’s comment at the end of page 3: the second step is highlighted as: “Then, a single hidden layer neural network was trained.”.
•	Figure 8 is updated with the revisions the author is recommended.
•	The authors agree that one machine learning is not necessarily superior to another in general. For this particular dataset one might have performed better than the other and we think it is useful information to provide quantitative analysis about it. The Markov chain provides a probabilistic framework that achieved better accuracy in comparison to only using a machine learning method. Based on the rules, the operation state (as described in fig. 8) and modeling results, the hidden Markov model finally makes a decision about the prediction. 

Thank you again,
Best,
Oney

**Comment by iodriller** [comment]
Hi @yingweiy,

Long time no talk, hope you are doing very well. Thanks for the review.

We are planning to do further research and publish more papers that will provide a more in-depth analysis of the rules and potentially multi-well analysis, as well. For now, as described in the present manuscript, simple rules are followed. Simple rules already provided a significant accuracy, one example from the text is: 

“For example, the SPP should be zero if drilling is temporarily halted and the drillstring is in slips with a zero flow rate.”. 

This particular example is quite general and can be applied across many wells/datasets without any calibration need. Another example of a simple rule was given as:

“For example, by understanding the operational state and historical SPP data, we can identify the inaccuracies of the physics-based model. This yields an advantage to the data-driven model. The hidden Markov model will then attribute a higher weight and trust to the data-driven model. Comparatively, the results from the data-driven model can be unrealistic for various reasons (i.e. outliers, sensor errors, etc.), and the physics model can be assigned a higher weight.”

As you recommended, I changed the order of the examples for better readability, so that some simple rules are clearer to the reader.

For this paper, we are presenting the analysis from a single well. It is a very good point to test the system via multi-well analysis, this can be within the scope of the future manuscripts.

The NN-based work considers the entire well dataset, while the GPs consider nearby stands.

Thanks again, Cheers,
Oney

**Comment by deniederhut** [comment]
Hi @ivan-marroquin and @yingweiy ! Are you happy with the updated manuscript?

**Comment by yingweiy** [comment]
> Hi @ivan-marroquin and @yingweiy ! Are you happy with the updated manuscript?

Hi @deniederhut,

Yes, I accept the updated manuscript.

Thanks,
Yingwei

**Comment by ivan-marroquin** [comment]
Hi @deniederhut 

I also accept the response and comments from the authors.

Thanks for all,

Ivan

File: papers/oney_erge/refs.bib [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+  author={Schreiber, Jacob},
+  journal={The Journal of Machine Learning Research},
+  volume={18},
+  number={1},
+  pages={5992--5997},
+  year={2017},
+  publisher={JMLR. org}
+}
+
+@article{oliphant2013python,
```
Comment (deniederhut) [inline papers/oney_erge/refs.bib:272]: This reference is missing a journal and volume number

**Comment by deniederhut** [comment]
@iodriller it looks like some of the references are missing DOIs. Could I ask you to add those in?

**Comment by iodriller** [comment]
Hi @deniederhut , I revised the references. Included DOIs (that I could find), added vol and journal name to oliphant2013python, improved them overall. Please let me know, if anything else is needed.

File: papers/oney_erge/oney_erge_paper.rst [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+Physics-based modeling of the Herschel and Bulkley fluid flow at each individual geometry was accomplished with the equations presented in the literature (:cite:`bourgoyne1991applied`, :cite:`aadnoy2009advanced`). For the flow in annuli, the equations from  (:cite:`erge2015effects`) were used, which consider the effects of drillpipe eccentricity within the wellbore and rotation of the drillpipe on frictional pressure losses. The physics-based equations are derived from the first principles and were coded in Python. An iterative numerical scheme was programmed for the Herschel and Bulkley fluid flow in pipes and annuli. SciPy’s scipy.optimize (:cite:`millman2011python`; :cite:`oliphant2006guide`) was used to solve for the turbulent flow friction factor. Numba  (:cite:`lam2015numba`) was used to accelerate the handling of computationally heavy functions.
```
Comment (cbcunc) [inline papers/oney_erge/oney_erge_paper.rst]: Please follow the citation guidelines for SciPy at https://www.scipy.org/citing.html

File: papers/oney_erge/oney_erge_paper.rst [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+3.2. Data-driven modeling
+-------------------------
+
+Deep learning neural networks perform very well in capturing the complex relationships of the data (:cite:`haykin1994neural`). A PyTorch (:cite:`paszke2019pytorch`) implementation of a single and multi-layered neural network was developed to learn from the drilling time-series sensory data. The network was trained with flow rate, rotation rate, bit depth and hole depth to predict the standpipe pressure. Before training the network, the data was preprocessed and transformed using Scikit-Learn’s preprocessing library (:cite:`pedregosa2011scikit`). And, the training and test datasets are converted into NumPy (:cite:`oliphant2013python`) arrays.
```
Comment (cbcunc) [inline papers/oney_erge/oney_erge_paper.rst]: Please follow the citation guidelines for NumPy at https://www.scipy.org/citing.html

**Review by cbcunc — CHANGES_REQUESTED** [review]
Minor updates to citations for SciPy and NumPy as specified by https://www.scipy.org/citing.html requested as part of final proceedings committee walk-through.

**Comment by iodriller** [comment]
Hello @cbcunc, thank you for pointing that out, I just corrected the citations for the NumPy and SciPy.

**Review by cbcunc — APPROVED** [review]
Thank you for your updates!
